src/extends/room/construction.js

```diff
@@ -319,3 +325,259 @@ class RoomLayout {
+        if (horizontal) {
+          let endpointOffset = exit === TOP ? 1 : -1
+          let y = piece[0].y
+          let left = new RoomPosition(piece[0].x - 1, y, this.roomname)
+          if (Game.map.getTerrainAt(piece[0].x - 1, y + endpointOffset) !== 'wall') {
+            map.set(piece[0].x - 1, y + endpointOffset, WALL_GATEWAY)
+          }
+          let lastPiece = piece.length - 1
+          let right = new RoomPosition(piece[lastPiece].x + 1, y, this.roomname)
+          if (Game.map.getTerrainAt(piece[lastPiece].x + 1, y + endpointOffset, this.roomname) !== 'wall') {
+            map.set(piece[lastPiece].x + 1, y + endpointOffset, WALL_GATEWAY)
+          }
+          piece.unshift(left)
+          piece.push(right)
+        } else {
+          let endpointOffset = exit === LEFT ? 1 : -1
+          let x = piece[0].x
+          let top = new RoomPosition(x, piece[0].y - 1, this.roomname)
+          if (Game.map.getTerrainAt(x + endpointOffset, piece[0].y - 1, this.roomname) !== 'wall') {
+            map.set(x + endpointOffset, piece[0].y - 1, WALL_GATEWAY)
+          }
+          let lastPiece = piece.length - 1
+          let bottom = new RoomPosition(x, piece[lastPiece].y + 1, this.roomname)
+          if (Game.map.getTerrainAt(x + endpointOffset, piece[lastPiece].y + 1, this.roomname) !== 'wall') {
+            map.set(x + endpointOffset, piece[lastPiece].y + 1, WALL_GATEWAY)
+          }
+          piece.unshift(top)
+          piece.push(bottom)
+        }
+
+        let type
+        for (let position of piece) {
+          type = type !== RAMPART_GATEWAY ? RAMPART_GATEWAY : WALL_GATEWAY
+          let x
+          let y
+          if (horizontal) {
+            x = position.x
+            y = position.y + (exit === TOP ? 2 : -2)
+          } else {
+            x = position.x + (exit === LEFT ? 2 : -2)
+            y = position.y
+          }
+          map.set(x, y, type)
+        }
+      }
+    }
+  }
+
+  _getExitPieces (exit) {
+    let vertical = exit === LEFT || exit === RIGHT
+    let stationary
+    if (vertical) {
+      stationary = exit === LEFT ? 0 : 49
+    } else {
+      stationary = exit === TOP ? 0 : 49
+    }
+
+    let pieces = []
+    let currentPiece = []
+    for (var i = 0; i <= 49; i++) {
+      const x = vertical ? stationary : i
+      const y = vertical ? i : stationary
+      const isExit = Game.map.getTerrainAt(x, y, this.roomname) !== 'wall'
+      if (isExit) {
+        currentPiece.push(new RoomPosition(x, y, this.roomname))
+      } else if (currentPiece.length > 0) {
+        pieces.push(currentPiece)
+        currentPiece = []
+      }
+    }
+    if (currentPiece.length > 0) {
+      pieces.push(currentPiece)
+    }
+    return pieces
+  }
+
+  // Using a room and a rampart map (CostMatrix) this generates a list of positions reachable from an exit.
+  _getWalkableMap (room, map) {
+    let tested = []
+    let queue = []
+    let walkableMap = new PathFinder.CostMatrix()
```

Comment:
Isn't it `new PathFinder.CostMatrix` Without the ()? I've seen that everywhere.
It looks off in those other locations, this looks better to me, I was just wondering if this works correctly.

Reply:
They're both equivalent to the same thing. From [mdn](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/new)-

> The constructor function Foo is called with the specified arguments, and with this bound to the newly created object. `new Foo` is equivalent to `new Foo()`, i.e. if no argument list is specified, Foo is called without arguments.

Basically you can drop the parenthesis as long as you aren't passing any arguments, but I think that looks gross.

Reply:
That's awesome. I really disliked the non-parenthesis  way.